File: Prototype/src/main/java/me/danetnaverno/editoni/editor/operations/SetBlocksOperation.java
package me.danetnaverno.editoni.editor.operations;

import me.danetnaverno.editoni.common.world.Block;
import me.danetnaverno.editoni.editor.Editor;

import java.util.Collection;
import java.util.stream.Collectors;

public class SetBlocksOperation extends Operation
{
    private Collection<Block> rollback;
    private Collection<Block> blocks;

    public SetBlocksOperation(Collection<Block> blocks)
    {
        this.blocks = blocks;
        rollback = blocks.stream().map( it-> Editor.INSTANCE.getWorld().getBlockAt(it.getGlobalX(), it.getGlobalY(), it.getGlobalZ())).collect(Collectors.toList());
    }

    public void apply()
    {
        innerApply(blocks);
    }

    public void rollback()
    {
        innerApply(rollback);
    }

    public void innerApply(Collection<Block> blocks)
    {
        for (Block block : blocks)
            Editor.INSTANCE.getWorld().setBlock(block);
    }
}

File: src/main/java/me/danetnaverno/editoni/operations/SetBlocksOperation.kt
package me.danetnaverno.editoni.operations

import me.danetnaverno.editoni.util.Translation
import me.danetnaverno.editoni.util.location.BlockArea
import me.danetnaverno.editoni.world.Block

open class SetBlocksOperation(area: BlockArea, protected val blocks: Collection<Block>) : Operation()
{
    override val displayName = Translation.translate("operation.set",
            blocks.minByOrNull { it.location.localX + it.location.localY + it.location.localZ },
            blocks.maxByOrNull { it.location.localX + it.location.localY + it.location.localZ })

    override val alteredChunks = area.toChunkArea()

    lateinit var replacedBlocks: Collection<Block>

    override fun initialApply()
    {
        replacedBlocks = blocks.mapNotNull { it.chunk.world.getLoadedBlockAt(it.location) }.toList()
        reapply()
    }

    override fun reapply()
    {
        val world = blocks.first().chunk.world
        for (block in blocks)
            world.setBlock(block)
    }

    override fun rollback()
    {
        val world = blocks.first().chunk.world
        for (oldBlock in replacedBlocks)
            world.setBlock(oldBlock)
    }
}